### ISSUE TITLE
Add raw_config_args option to providers

### ISSUE THREAD

**Issue by mjuenema** [metacloud/molecule#504]
Currently only the `vagrant` section supports `raw_config_args`. I am currently trialling changing `vagrantfile.j2` so that the `providers` section also allows `raw_config_args`.

Example:

```
providers:
  - name: virtualbox
    type: virtualbox
    options:
      memory: 512
    raw_config_args:
      - "customize \"post-boot\",[\"controlvm\", :id, \"setlinkstate1\", \"on\"]"
```

I am going to submit a pull request once I am confident that I got it all working. Any thoughts on this?

**Comment by signed8bit** [comment]
Hi @mjuenema, I am currently working on moving us away from the Jinja2 template for the Vagrantfile in #463. You might want to wait for that to land. Regardless, supporting raw config args in the providers section seems reasonable to me.

**Comment by retr0h** [comment]
We actually already support [this](https://github.com/metacloud/molecule/issues/424).  As we start to move towards a more test kitchen like config file, we will move this to the provider section of the config.  However, as of today, you should be able to do what you described.

**Comment by retr0h** [comment]
You're right.

**Comment by abrown-sg** [comment]
Another issue we have is that we can't set provider options at the instance level. So, for example, changing memory per instance or attaching a disk image per instance.  Vagrantfile supports letting you change provider options from within an instance in multi-machine scenarios like we have now.  For example, here's a workaround we're using:

vagrantfile.j2
...snip...
```
{% for instance in config.vagrant.instances -%}
  config.vm.define '{{ instance.vm_name }}' do |c|
    {%- if 'libvirt' not in config.vagrant.providers|map(attribute='type')|list %}
    c.vm.hostname = '{{ instance.vm_name }}'
    {%- endif %}
    {% if instance.platform %}
    c.vm.box = '{{ instance.platform }}'
    {% endif %}
    {%- for interface in instance.interfaces %}
    c.vm.network '{{ interface.network_name }}', type: '{{ interface.type }}', auto_config: {{ interface.auto_config|lower }}
    {%- if interface.type == 'static' %}, ip: '{{ interface.ip }}' {%- endif -%}
    {%- endfor %}
    {%- if 'provider_args' in instance %}
    c.vm.provider 'virtualbox' do |vb|
      {%- for line in instance.provider_args %}
      vb.{{ line }}
      {%- endfor %}
    end
    {%- endif %}
    {%- if 'raw_config_args' in instance %}
      {%- for line in instance.raw_config_args %}
    c.{{ line }}
      {%- endfor %}
    {%- endif %}
  end
  {% endfor %}
end
```

Then, in molecule.yml we define something like:
```
molecule:
  vagrantfile_template: ./vagrantfile.j2

vagrant:
  instances:
    - name: test-01
      provider_args:
        - "customize ['createhd', '--filename', 'ssd.vdi', '--size', 5 * 1024]"
        - "customize ['storageattach', :id, '--storagectl', 'SATA', '--port', 1, '--device', 0, '--type', 'hdd', '--medium', 'ssd.vdi']"
```

And the rendered .molecule/vagrantfile will look like:
```
Vagrant.configure('2') do |config|
  config.cache.scope = 'machine' if Vagrant.has_plugin?('vagrant-cachier')
  config.vm.box = '...'
  config.vm.box_url = '...'

  config.vm.provider 'virtualbox' do |vb|
    vb.memory = 1024
    vb.cpus = 2
    vb.linked_clone = true if Vagrant::VERSION =~ /^1.8/
  end

  config.vm.define 'test-01' do |c|
    c.vm.hostname = 'test-01'

    c.vm.network 'private_network', type: 'dhcp', auto_config: true
    c.vm.provider 'virtualbox' do |vb|
      vb.customize ['createhd', '--filename', 'ssd.vdi', '--size', 5 * 1024]
      vb.customize ['storageattach', :id, '--storagectl', 'SATA', '--port', 1, '--device', 0, '--type', 'hdd', '--medium', 'ssd.vdi']
    end
  end
end
```

This was just a quick/dirty hack that only supports virtualbox currently, but it appears to be the correct way forward when trying to set provider specific arguments at the instance level.

**Comment by bbaassssiiee** [comment]
I would like to use raw_config_args in ~/.config/molecule/config.yml

**Comment by retr0h** [comment]
> This was just a quick/dirty hack that only supports virtualbox currently, but it appears to be the correct way forward when trying to set provider specific arguments at the instance level.

It should be noted the above "hack" cannot be accomplished with molecule any longer.  We no longer support the ability to provide your own Vagrantfile template.  Since the Vagrantfile is coupled so tightly with molecule, allowing one to provided their own would only cause issues, in the form of GH issues.

**Comment by abrown-sg** [comment]
Is that true? It looks like you can still do it.

```
$ molecule create --debug |grep vagrantfile_template
      vagrantfile_template: ./foo.j2

$ head -3 molecule.yml
---
molecule:
  vagrantfile_template: ./foo.j2

$ molecule --version
1.11.4
```

**Comment by retr0h** [comment]
> I would like to use raw_config_args in ~/.config/molecule/config.yml

Yeah, I know the problem with this one.  This is due to the way in which the config is merged, and the decision to make `providers` a list vs a `dict`, so it could get properly merged.  This will require thought in general, and a known issue.  I'm not sure when we will address it.  Would be slated for fixup with #204.

I'm not entirely sure how I want to fix this issue in general.  The implementation of `raw_config_args` needs reworked, so it is provider agnostic.  As it stands today (as @mjuenema mentioned), changing the provider could break the options passed to `raw_config_args`.

This will be fixed sometime after #463 is implemented.

**Comment by retr0h** [comment]
> Is that true? It looks like you can still do it.

You know what, you're right.  However, it will be removed once #463 lands, since templates will be moved under [cookiecutter](http://cookiecutter.readthedocs.io/en/latest/) ownership.

**Comment by retr0h** [comment]
Should be corrected with #687.

**Comment by signed8bit** [comment]
There is actually some additional changes beyond #687 for this to work as stated. The previous work I did just brought over what the old .j2 file supported. I will get a PR for this today though.

**Comment by retr0h** [comment]
Corrected in #724.
